lib/portlayer/exec/container.go

```diff
@@ -280,69 +301,65 @@ func (c *Container) Remove(ctx context.Context, sess *session.Session) error {
+func (c *Container) Update(ctx context.Context, sess *session.Session) (*Handle, error) {
 	defer trace.End(trace.Begin("Container.Update"))
 	c.Lock()
 	defer c.Unlock()
 
 	if c.vm == nil {
-		return nil, fmt.Errorf("container does not have a vm")
-	}
+		vm, err := childVM(ctx, sess, c.ExecConfig.ID)
+		if err != nil {
+			return nil, err
+		}
 
-	var vm []mo.VirtualMachine
+		c.vm = vm
+	}
 
-	if err := sess.Retrieve(ctx, []types.ManagedObjectReference{c.vm.Reference()}, []string{"config"}, &vm); err != nil {
+	if err := c.info(ctx, sess); err != nil {
 		return nil, err
 	}
 
-	extraconfig.Decode(vmomi.OptionValueSource(vm[0].Config.ExtraConfig), c.ExecConfig)
-	return c.ExecConfig, nil
+	h := c.newHandle()
+	h.State = new(State)
+	*h.State = c.State
+	return h, nil
 }
 
-// Grab the info for the requested container
-// TODO:  Possibly change so that handler requests a handle to the
-// container and if it's not present then search and return a handle
-func ContainerInfo(ctx context.Context, sess *session.Session, containerID ID) (*Container, error) {
-	// first  lets see if we have it in the cache
-	container := containers.Container(containerID.String())
-	// if we missed search for it...
-	if container == nil {
-		// search
-		vm, err := childVM(ctx, sess, containerID.String())
-		if err != nil || vm == nil {
-			log.Debugf("ContainerInfo failed to find childVM: %s", err.Error())
-			return nil, fmt.Errorf("Container Not Found: %s", containerID)
-		}
-		container = &Container{vm: vm}
-	}
+func (c *Container) swap(other *Container) {
+	c.ExecConfig, other.ExecConfig = other.ExecConfig, c.ExecConfig
+	c.State, other.State = other.State, c.State
+	c.VMUnsharedDisk, other.VMUnsharedDisk = other.VMUnsharedDisk, c.VMUnsharedDisk
+	c.vm, other.vm = other.vm, c.vm
+}
 
+func (c *Container) info(ctx context.Context, sess *session.Session) error {
 	// get properties for specific containerVMs
 	// we must do this since we don't know that we have a valid state
 	// This will be refactored when event streaming hits
-	vms, err := populateVMAttributes(ctx, sess, []types.ManagedObjectReference{container.vm.Reference()})
+	vms, err := populateVMAttributes(ctx, sess, []types.ManagedObjectReference{c.vm.Reference()})
 	if err != nil {
-		return nil, err
+		return err
 	}
 	// convert the VMs to container objects -- include
 	// powered off vms
-	all := true
-	cc := convertInfraContainers(vms, &all)
+	cc := convertInfraContainers(vms, true)
 
 	switch len(cc) {
 	case 0:
 		// we found a vm, but it doesn't appear to be a container VM
-		return nil, fmt.Errorf("%s does not appear to be a container", containerID)
+		return fmt.Errorf("not a container")
 	case 1:
 		// we have a winner
-		return cc[0], nil
-	default:
-		// we manged to find multiple vms
-		return nil, fmt.Errorf("multiple containers named %s found", containerID)
+		cc[0].vm = c.vm
+		c.swap(cc[0])
```

Comment:
why do you need this swap? Is this to update the cached container information based on VM information?